### ISSUE TITLE
indexing of a subarray,(bug?)

### ISSUE THREAD

**Issue by pingzou** [JuliaLang/julia#14509]
I create a subarray, then index into it:
```
In[1]:   a=randn(5,5,2,2);
           b=sub(a,5,:,:,:)
Out[1] :
1x5x2x2 SubArray{Float64,4,Array{Float64,4},Tuple{UnitRange{Int64},Colon,Colon,Colon},4}:
[:, :, 1, 1] =
 -0.377684  0.440743  0.589393  0.129113  -1.0641

[:, :, 2, 1] =
 0.457265  1.19336  0.381148  -1.27136  -0.43993

[:, :, 1, 2] =
 0.609543  -1.31754  1.81003  -0.437679  -1.7368

[:, :, 2, 2] =
 -0.260656  0.540009  0.194623  -1.20884  0.248411


In[2]:      b[1:end,:,1,1]
Out[2]:    1x5 Array{Float64,2}:
 -0.377684  0.440743  0.589393  0.129113  -1.0641


In[3]:     b[:,:,1,1]
Out[3]:   1x5 Array{Float64,2}:
 -0.377684  -1.10236  2.75113  -1.80442  -0.234724
```
The result  of `b[:,:,1,1]` is different from  `b[1:end,:,1,1]=b[1,:,1,1]`,
and it seems strange.
when  indexing of `sub(a,:,5,:,:)`,or `sub(a,4:5,:,:,:)`, the results are as expected.

(The Julia version : 0.4.2, os: ubuntu15.10)

**Comment by andreasnoack** [comment]
For some reason the `LD` parameter is not determined correctly when slicing
```
julia> slice(b, :,:,1,1)
1x5 SubArray{Float64,2,Array{Float64,4},Tuple{UnitRange{Int64},Colon,Int64,Int64},4}:
 0.403486  -0.899161  -0.280601  -1.14374  -0.462667

julia> slice(b, 1:1,:,1,1)
1x5 SubArray{Float64,2,Array{Float64,4},Tuple{UnitRange{Int64},Colon,Int64,Int64},1}:
 0.403486  -0.899161  -0.280601  -1.14374  -0.462667
```
and therefore `Base.linearindexing` wrongly returns `LinearFast()`.

cc: @timholy, @mbauman

**Comment by timholy** [comment]
@pingzou, thanks for finding and reporting this awful bug.

I'm working on it, but it's quite tricky---every fix I've tried so far has broken something else. In those examples, the `LD` parameter is technically correct, it's the `stride1` that's wrong. The thing that makes this hard is that with `sub` we convert `5` into `5:5` and then don't keep a record that it was once `5`, except from some hints in the `LD` parameter that don't specify the individual index slot(s). We make use of those hints when we create a view-of-a-view and calculate the new `LD`, but this may not be sufficient to correctly update the `stride1` computation.

Simplified example:
```jl
julia> a = reshape(1:27, 3, 3, 3)
3x3x3 Array{Int64,3}:
[:, :, 1] =
 1  4  7
 2  5  8
 3  6  9

[:, :, 2] =
 10  13  16
 11  14  17
 12  15  18

[:, :, 3] =
 19  22  25
 20  23  26
 21  24  27

julia> b1 = sub(a, 2, 1:2, :)
1x2x3 SubArray{Int64,3,Array{Int64,3},Tuple{UnitRange{Int64},UnitRange{Int64},Colon},2}:
[:, :, 1] =
 2  5

[:, :, 2] =
 11  14

[:, :, 3] =
 20  23

julia> b2 = sub(a, 1:2, 2, :)
2x1x3 SubArray{Int64,3,Array{Int64,3},Tuple{UnitRange{Int64},UnitRange{Int64},Colon},2}:
[:, :, 1] =
 4
 5

[:, :, 2] =
 13
 14

[:, :, 3] =
 22
 23

julia> b1a = sub(b1, :, :, 2)   # note the LD parameter "correctly" goes from 2 to 3
1x2 SubArray{Int64,2,Array{Int64,3},Tuple{UnitRange{Int64},UnitRange{Int64},Int64},3}:
 11  14

julia> b2a = sub(b2, :, :, 2)   # here also
2x1 SubArray{Int64,2,Array{Int64,3},Tuple{UnitRange{Int64},UnitRange{Int64},Int64},3}:
 13
 14

julia> b1a.stride1     # but this one is wrong
1

julia> b2a.stride1
1
```
None of this is a problem with `slice`:
```jl
julia> b1 = slice(a, 2, 1:2, :)
2x3 SubArray{Int64,2,Array{Int64,3},Tuple{Int64,UnitRange{Int64},Colon},2}:
 2  11  20
 5  14  23

julia> b2 = slice(a, 1:2, 2, :)
2x3 SubArray{Int64,2,Array{Int64,3},Tuple{UnitRange{Int64},Int64,Colon},2}:
 4  13  22
 5  14  23

julia> b1a = sub(b1, :, 2)
2-element SubArray{Int64,1,Array{Int64,3},Tuple{Int64,UnitRange{Int64},Int64},3}:
 11
 14

julia> b2a = sub(b2, :, 2)
2-element SubArray{Int64,1,Array{Int64,3},Tuple{UnitRange{Int64},Int64,Int64},3}:
 13
 14

julia> b1a.stride1
3

julia> b2a.stride1
1
```

Currently I'm still trying to determine whether it's possible to compute `stride1` correctly given the information available. If not, the two options are (1) add a field to `SubArray` that stores the necessary information (this might be bad to do in the middle of a stable release, for e.g., JLD and serializer backward compatibility) (2) being explicitly pessimistic about `LD` (never returning an `LD` bigger than the parent array), but this will lead to performance regressions.

**Comment by timholy** [comment]
@pingzou, fixed in master. Hopefully the fix will also be backported in julia 0.4.3.
